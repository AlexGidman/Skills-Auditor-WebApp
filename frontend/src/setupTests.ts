// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
