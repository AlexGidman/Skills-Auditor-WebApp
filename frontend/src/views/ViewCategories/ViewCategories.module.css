.buttonWrapper {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    gap: var(--s2);
    margin: 0 10%;
}

.button {
    flex: 40%;
}
.header {
    margin-top: 10%;
}

.welcomeMessage {
    font-size: xx-large;
}

.listview {
    margin-bottom: var(--s2);
    margin-left: 10%;
    margin-right: 10%;
}

@media (max-width: 900px) {
    .buttonWrapper {
        margin: 0;
    }

    .listview {
        margin-left: 0;
        margin-right: 0;
    }
}
