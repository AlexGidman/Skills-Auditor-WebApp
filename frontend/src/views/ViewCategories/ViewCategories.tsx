import React, { useState, useEffect } from "react";
import { LoadingPlaceholder } from "../../components/LoadingPlaceholder/LoadingPlaceholder";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { LinkButton } from "../../components/LinkButton/LinkButton";
import { Button } from "../../components/Button/Button";
import { getSelectOptionsFromArray, useAPI, useIsAdminOrManager } from "../../utility/helper";
import { getAllCategories, deleteCategory } from "../../utility/apiRequests";
import styles from "./ViewCategories.module.css";
import { ListView } from "../../components/ListView/ListView";
import { useOutletContext, useNavigate } from "react-router-dom";
import { Category } from "../../utility/types";

export const ViewCategories = () => {
    useIsAdminOrManager();
    // @ts-ignore TODO fix type for AppOutletContext here
    const [, setShowToast] = useOutletContext();
    const navigate = useNavigate();
    const [selectedCategory, setSelectedCategory] = useState<string>();
    const [refresh, setRefresh] = useState(false);

    const { data, loading, error } = useAPI<Category[]>(getAllCategories);
    const {
        data: deleteCategoryData,
        loading: deleteCategoryLoading,
        error: deleteCategoryError,
        callback: executeDeleteCategory,
    } = useAPI(deleteCategory, [selectedCategory], true);

    useEffect(() => {
        if (refresh && !deleteCategoryLoading) {
            setRefresh(false);
            deleteCategoryError && setShowToast({ error: "Failed to delete category" });
            deleteCategoryData && navigate(0);
        }
    }, [
        deleteCategoryLoading,
        deleteCategoryData,
        deleteCategoryError,
        navigate,
        refresh,
        setShowToast,
    ]);

    if (loading) return <LoadingPlaceholder />;
    if (error) return <ErrorMessage error={error} />;
    if (data) {
        const sortedCategories = data.sort((a, b) => a.name.localeCompare(b.name));
        return (
            <>
                <header>
                    <h1>View Categories</h1>
                </header>
                <section>
                    <ListView
                        className={styles.listview}
                        options={getSelectOptionsFromArray(
                            sortedCategories.map((category) => category.id),
                            sortedCategories.map((category) => category.name),
                        )}
                        value={selectedCategory}
                        labelText="Select a category"
                        onChange={(e) => {
                            setSelectedCategory(e.target.value);
                        }}
                    />
                    <div className={styles.buttonWrapper}>
                        <LinkButton
                            className={styles.button}
                            disabled={!selectedCategory}
                            name="Edit category"
                            path={`/edit/category/${selectedCategory}`}
                        />
                        <Button
                            className={styles.button}
                            loading={loading || deleteCategoryLoading}
                            disabled={!selectedCategory}
                            onClick={(e) => {
                                e.preventDefault();
                                executeDeleteCategory();
                                setRefresh(true);
                            }}
                        >
                            Delete Category
                        </Button>
                    </div>
                </section>
            </>
        );
    }
    return null;
};
