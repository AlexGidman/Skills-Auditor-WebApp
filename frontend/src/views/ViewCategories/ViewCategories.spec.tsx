import { render, screen } from "@testing-library/react";
import { ViewCategories } from "./ViewCategories";
import { mockAdminUser, mockApiRequests, mockError } from "../../setupTests";
import userEvent from "@testing-library/user-event";
import * as ReactDom from "react-router-dom";
import { wait } from "@testing-library/user-event/dist/utils";
import React from "react";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
    };
});

const mockCategory1 = {
    id: 1,
    name: "Category1",
};

const mockCategory2 = {
    id: 2,
    name: "Category2",
};

describe("ViewCategories", () => {
    it("renders correctly when data successful", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue([mockAdminUser]);
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllCategories.mockImplementation(() =>
            Promise.resolve({ data: [mockCategory1, mockCategory2] }),
        );
        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewCategories />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Categories" })).toBeInTheDocument();

        expect(screen.getByRole("listbox")).toBeInTheDocument();
        expect(screen.getByRole("option", { name: mockCategory1.name })).toBeInTheDocument();
        expect(screen.getByRole("option", { name: mockCategory2.name })).toBeInTheDocument();

        expect(screen.getByRole("link", { name: "Edit category" })).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it("should have Edit category link that has no valid href if no option is selected", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue([mockAdminUser]);
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllCategories.mockImplementation(() =>
            Promise.resolve({ data: [mockCategory1, mockCategory2] }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewCategories />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Categories" })).toBeInTheDocument();

        const options = screen.getAllByRole("option");
        options.forEach((option) => {
            expect(option).not.toBeChecked();
        });

        expect(screen.getByRole("link", { name: "Edit category" })).toHaveAttribute(
            "href",
            "/edit/category/undefined",
        );
    });

    it("should have Edit category link that has a valid href if option is selected", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue([mockAdminUser]);
        // @ts-ignore TODO: fix this
        mockApiRequests.getAllCategories.mockImplementation(() =>
            Promise.resolve({ data: [mockCategory1, mockCategory2] }),
        );
        render(
            <ReactDom.MemoryRouter>
                <ViewCategories />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "View Categories" })).toBeInTheDocument();

        userEvent.selectOptions(screen.getByRole("listbox"), "1");
        await wait(100);

        expect(screen.getByRole("link", { name: "Edit category" })).toHaveAttribute(
            "href",
            `/edit/category/${mockCategory1.id}`,
        );
    });

    it("renders correctly when error", async () => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue([mockAdminUser]);
        mockApiRequests.getAllCategories.mockImplementation(() => Promise.reject(mockError));
        const { container } = render(
            <ReactDom.MemoryRouter>
                <ViewCategories />
            </ReactDom.MemoryRouter>,
        );
        expect(
            await screen.findByRole("alert", { name: mockError.response.data.error.message }),
        ).toBeInTheDocument();

        expect(screen.queryByRole("listbox")).not.toBeInTheDocument();

        expect(screen.queryByRole("link", { name: "Edit category" })).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });
});
