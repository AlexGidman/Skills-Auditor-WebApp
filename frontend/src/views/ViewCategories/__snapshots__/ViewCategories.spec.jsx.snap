// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewCategories renders correctly when data successful 1`] = `
<div>
  <header>
    <h1>
      View Categories
    </h1>
  </header>
  <section>
    <div
      class="component component listview"
    >
      <label
        class="label"
        for="Select-a category-undefined"
      >
        Select a category
        :
      </label>
      <select
        aria-label="Select a category"
        class="select"
        name="Select a category"
        size="5"
      >
        <option
          value="1"
        >
          Category1
        </option>
        <option
          value="2"
        >
          Category2
        </option>
      </select>
    </div>
    <div
      class="buttonWrapper"
    >
      <a
        class="component button disabled"
        href="/edit/category/undefined"
      >
        Edit category
      </a>
      <button
        class="component button"
        disabled=""
      >
        Delete Category
      </button>
    </div>
  </section>
</div>
`;

exports[`ViewCategories renders correctly when error 1`] = `
<div>
  <div
    aria-label="Oops"
    class="component"
    role="alert"
  >
    <h1>
      Something went wrong...
    </h1>
    <p>
      Oops
    </p>
  </div>
</div>
`;
